### PR TITLE
Fix auto response options

### PR DESCRIPTION
**Pull request by rustamwin**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ✔️/❌
| Fixed issues  | #136

### CODE REVIEW

**Comment by samdark** [comment]
https://github.com/yiisoft/router/issues/136

File: src/Middleware/Router.php [hunk 1]
```diff
@@ -43,7 +43,7 @@ public function process(ServerRequestInterface $request, RequestHandlerInterface
         $this->currentRoute->setUri($request->getUri());
 
         if ($result->isMethodFailure()) {
-            if ($this->autoResponseOptions && $request->getMethod() === Method::OPTIONS) {
+            if ($request->getMethod() === Method::OPTIONS && $this->isSameOrigin($request)) {
```
Comment (yiiliveext) [inline src/Middleware/Router.php:46]: We must generate this response for any origin.

File: src/Group.php [hunk 2]
```diff
@@ -87,6 +88,34 @@ public function withDispatcher(MiddlewareDispatcher $dispatcher): GroupInterface
...
+        }
+        $group = clone $this;
+        $pattern = '';
+        foreach ($group->items as $index => $item) {
+            if ($item instanceof self) {
+                $item = $item->withAutoOptions(...$middlewares);
+                $group->items[$index] = $item;
+            } else {
+                // Avoid duplicates
+                if ($pattern === $item->getPattern() || in_array(Method::OPTIONS, $item->getMethods(), true)) {
```
Comment (yiiliveext) [inline src/Group.php:104]: What about a host?

Comment (yiiliveext) [reply src/Group.php:104]: Also, this method doesn't allow to avoid all duplicates.

File: src/Group.php [hunk 3]
```diff
@@ -87,6 +90,38 @@ public function withDispatcher(MiddlewareDispatcher $dispatcher): GroupInterface
...
+                    continue;
+                }
+                $pattern = $item->getPattern();
+                $route = Route::options($pattern);
+                foreach ($middlewares as $middleware) {
+                    $route = $route->middleware($middleware);
+                }
+                $group->items[] = $route->action(
+                    static fn (ServerRequestInterface $request, RequestHandlerInterface $handler) => $handler->handle(
```
Comment (yiiliveext) [inline src/Group.php:115]: The next handler is NotFoundHandler

**Comment by samdark** [comment]
Fix confirmed by @acidka who reported original issue in the chat.

File: src/Group.php [hunk 4]
```diff
@@ -87,6 +88,24 @@ public function withDispatcher(MiddlewareDispatcher $dispatcher): GroupInterface
         return $group;
     }
 
+    public function withAutoOptions(...$middlewares): GroupInterface
```
Comment (samdark) [inline src/Group.php:91]: Need phpdoc.

File: src/Group.php [hunk 5]
```diff
@@ -87,6 +88,24 @@ public function withDispatcher(MiddlewareDispatcher $dispatcher): GroupInterface
...
+    {
+        $group = clone $this;
+        $group->autoOptions = array_merge($group->autoOptions, $middlewares);
+
+        return $group;
+    }
+
+    public function getAutoOptions(): array
```
Comment (samdark) [inline src/Group.php:99]: Need phpdoc.

File: src/Group.php [hunk 6]
```diff
@@ -87,6 +88,24 @@ public function withDispatcher(MiddlewareDispatcher $dispatcher): GroupInterface
...
+    {
+        return $this->autoOptions;
+    }
+
+    public function hasAutoOptions(): bool
```
Comment (samdark) [inline src/Group.php:104]: Need phpdoc.

File: src/GroupInterface.php [hunk 7]
```diff
@@ -46,4 +46,6 @@ public function host(string $host): self;
     public function namePrefix(string $namePrefix): self;
 
     public function routes(...$routes): self;
+
+    public function withAutoOptions(...$middlewares): self;
```
Comment (samdark) [inline src/GroupInterface.php:50]: Need phpdoc.

File: src/Group.php (hunk 4 above)
Comment (samdark) [reply src/Group.php:91]: Maybe it is worth naming it `withAutoOptionsMiddlewares()`. It is a bit longer but makes user to wonder less about the method purpose.

File: src/Group.php (hunk 5 above)
Comment (samdark) [reply src/Group.php:99]: `getAutoOptionsMiddlewares()`?

File: src/Group.php (hunk 6 above)
Comment (samdark) [reply src/Group.php:104]: `hasAutoOptionsMiddlewares()`?

File: src/GroupInterface.php (hunk 7 above)
Comment (samdark) [reply src/GroupInterface.php:50]: `withAutoOptionsMiddlewares()`?

File: src/RouteCollection.php [hunk 8]
```diff
@@ -139,6 +149,37 @@ private function injectGroup(Group $group, array &$tree, string $prefix = '', st
         }
     }
 
+    private function processAutoOptions($group, &$host, &$pattern, &$modifiedItem, &$tree): void
+    {
+        if (
+            $group->hasAutoOptions()
+            && !(($pattern === $modifiedItem->getPattern() && $host === $modifiedItem->getHost())
+                || in_array(Method::OPTIONS, $modifiedItem->getMethods(), true))
+        ) {
```
Comment (samdark) [inline src/RouteCollection.php:158]:
```suggestion
        if (
            $group->hasAutoOptions()
            && !in_array(Method::OPTIONS, $modifiedItem->getMethods(), true)
            && ($pattern !== $modifiedItem->getPattern() || $host !== $modifiedItem->getHost())
        ) {
```

File: src/RouteCollection.php [hunk 9]
```diff
@@ -139,6 +149,37 @@ private function injectGroup(Group $group, array &$tree, string $prefix = '', st
...
+            $pattern = $modifiedItem->getPattern();
+            $host = $modifiedItem->getHost();
```
Comment (samdark) [inline src/RouteCollection.php:160]: These two could be moved out of `if`.

**Review by samdark — CHANGES_REQUESTED** [review]
The solution itself is confirmed to solve the issue. Some changes are required though. Also, README additions are needed.

File: src/GroupInterface.php (hunk 7 above)
Comment (yiiliveext) [reply src/GroupInterface.php:50]: No, this method is not only for middlewares, it appends OPTIONS routes.

Comment (samdark) [reply src/GroupInterface.php:50]: OK.

**Comment by samdark** [comment]
https://github.com/yiisoft/router/pull/137

File: src/Group.php [hunk 10]
```diff
@@ -25,7 +25,7 @@ final class Group implements GroupInterface
     private bool $routesAdded = false;
     private bool $middlewareAdded = false;
     private array $disabledMiddlewareDefinitions = [];
-    private array $autoOptions = [];
+    private array $corsMiddlewares = [];
```
Comment (samdark) [inline src/Group.php:28]: Do we need an array if `withCors()` now accepts a single middleware?

Comment (rustamwin) [reply src/Group.php:28]: Yes `withCors()` now accepts a single middleware definition but, can be used multiple times `withCors()->withCors()`...

Comment (acidka) [reply src/Group.php:28]: Any scenario there more than one CORS middleware needed? just asking

File: src/RouteCollection.php (hunk 9 above)
Comment (rustamwin) [reply src/RouteCollection.php:176]: No, it breaks the code.

File: src/Group.php (hunk 10 above)
Comment (samdark) [reply src/Group.php:28]: I don't have any.

**Comment by samdark** [comment]
Merged. Thanks for handling it, @rustamwin, @yiiliveext.

**Pull request merged**